File: src/core/continuation_capsule.zig
//! Proof-carrying continuation manifest for one committed AI checkpoint.
//!
//! The fixed-size wire binds nine typed external objects without copying their
//! payloads. Verification receives the expected scalar identity and exact
//! object bytes, recomputes every domain-separated reference, and rejects any
//! substitution. The manifest grants no model, allocator, scheduler, file, or
//! output authority; a resolver must supply those capabilities separately.

const std = @import("std");

pub const Digest = [32]u8;
pub const zero_digest: Digest = [_]u8{0} ** 32;
pub const wire_abi: u64 = 0x4743_4341_0000_0001;
pub const magic = [_]u8{ 'G', 'C', 'C', 'A', 'P', 'V', '0', '1' };
pub const flag_require_all_objects: u32 = 1 << 0;
pub const allowed_flags: u32 = flag_require_all_objects;

const object_domain = "glacier-continuation-object-v1\x00";
const envelope_domain = "glacier-continuation-capsule-wire-v1\x00";
const digest_bytes = @sizeOf(Digest);

pub const ObjectKind = enum(u64) {
    model = 0,
    tokenizer = 1,
    execution_plan = 2,
    resource_state = 3,
    lane_state = 4,
    kv_state = 5,
    sampler_state = 6,
    output_state = 7,
    publication_receipt = 8,
};

pub const object_kinds = [_]ObjectKind{
    .model,
    .tokenizer,
    .execution_plan,
    .resource_state,
    .lane_state,
    .kv_state,
    .sampler_state,
    .output_state,
    .publication_receipt,
};

pub const object_count: usize = object_kinds.len;
pub const header_bytes: usize =
    magic.len + 8 + 8 + 4 + 4 + 6 * 8 + 2 * digest_bytes;
pub const object_ref_bytes: usize = 8 + 8 + digest_bytes;
pub const encoded_bytes: usize =
    header_bytes + object_count * object_ref_bytes + digest_bytes;

pub const Error = error{
    CapacityExceeded,
    InvalidStorage,
    InvalidMagic,
    InvalidAbi,
    InvalidLength,
    InvalidFlags,
    InvalidManifest,
    InvalidObject,
    InvalidEnvelope,
    InvalidComposition,
};

pub const ObjectV1 = struct {
    abi_version: u64,
    bytes: []const u8,
};

pub const ObjectsV1 = struct {
    model: ObjectV1,
    tokenizer: ObjectV1,
    execution_plan: ObjectV1,
    resource_state: ObjectV1,
    lane_state: ObjectV1,
    kv_state: ObjectV1,
    sampler_state: ObjectV1,
    output_state: ObjectV1,
    publication_receipt: ObjectV1,

    pub fn get(self: ObjectsV1, kind: ObjectKind) ObjectV1 {
        return switch (kind) {
            .model => self.model,
            .tokenizer => self.tokenizer,
            .execution_plan => self.execution_plan,
            .resource_state => self.resource_state,
            .lane_state => self.lane_state,
            .kv_state => self.kv_state,
            .sampler_state => self.sampler_state,
            .output_state => self.output_state,
            .publication_receipt => self.publication_receipt,
        };
    }
};

pub const ConfigV1 = struct {
    execution_abi: u64,
    request_epoch: u64,
    publication_sequence: u64,
    checkpoint_generation: u64,
    kv_tokens: u64,
    output_tokens: u64,
    challenge_sha256: Digest,
    parent_capsule_sha256: Digest = zero_digest,
};

pub const ObjectRefV1 = struct {
    abi_version: u64,
    byte_length: u64,
    sha256: Digest,
};

pub const DecodedV1 = struct {
    config: ConfigV1,
    refs: [object_count]ObjectRefV1,
    envelope_sha256: Digest,

    pub fn ref(self: DecodedV1, kind: ObjectKind) ObjectRefV1 {
        return self.refs[@intCast(@intFromEnum(kind))];
    }
};

pub fn encodeV1(
    config: ConfigV1,
    objects: ObjectsV1,
    destination: []u8,
) Error![]const u8 {
    if (destination.len < encoded_bytes) return Error.CapacityExceeded;
    const output = destination[0..encoded_bytes];
    try validateConfig(config);
    for (object_kinds) |kind| {
        const object = objects.get(kind);
        try validateObject(object);
        if (slicesOverlap(output, object.bytes))
            return Error.InvalidStorage;
    }

    @memset(output, 0);
    errdefer @memset(output, 0);
    var writer: Writer = .{ .bytes = output };
    try writer.writeBytes(&magic);
    try writer.writeU64(wire_abi);
    try writer.writeU64(encoded_bytes);
    try writer.writeU32(flag_require_all_objects);
    try writer.writeU32(0);
    try writer.writeU64(config.execution_abi);
    try writer.writeU64(config.request_epoch);
    try writer.writeU64(config.publication_sequence);
    try writer.writeU64(config.checkpoint_generation);
    try writer.writeU64(config.kv_tokens);
    try writer.writeU64(config.output_tokens);
    try writer.writeDigest(config.challenge_sha256);
    try writer.writeDigest(config.parent_capsule_sha256);

    for (object_kinds) |kind| {
        const object_ref = try objectRefV1(kind, objects.get(kind));
        try writer.writeU64(object_ref.abi_version);
        try writer.writeU64(object_ref.byte_length);
        try writer.writeDigest(object_ref.sha256);
    }
    if (writer.position + digest_bytes != output.len)
        return Error.InvalidLength;
    try writer.writeDigest(envelopeSha256(output[0..writer.position]));
    if (writer.position != output.len) return Error.InvalidLength;
    return output;
}

/// Parse and validate the self-contained manifest envelope.
///
/// This does not verify external object payloads. Use `decodeAndVerifyV1` at a
/// resume/admission boundary before granting any authority.
pub fn decodeManifestV1(encoded: []const u8) Error!DecodedV1 {
    if (encoded.len != encoded_bytes) return Error.InvalidLength;
    var reader: Reader = .{ .bytes = encoded };
    if (!std.mem.eql(u8, try reader.readBytes(magic.len), &magic))
        return Error.InvalidMagic;
    if (try reader.readU64() != wire_abi) return Error.InvalidAbi;
    if (try reader.readU64() != encoded_bytes) return Error.InvalidLength;
    if (try reader.readU32() != flag_require_all_objects or
        try reader.readU32() != 0)
        return Error.InvalidFlags;

    const config: ConfigV1 = .{
        .execution_abi = try reader.readU64(),
        .request_epoch = try reader.readU64(),
        .publication_sequence = try reader.readU64(),
        .checkpoint_generation = try reader.readU64(),
        .kv_tokens = try reader.readU64(),
        .output_tokens = try reader.readU64(),
        .challenge_sha256 = try reader.readDigest(),
        .parent_capsule_sha256 = try reader.readDigest(),
    };
    try validateConfig(config);

    var refs: [object_count]ObjectRefV1 = undefined;
    for (&refs) |*object_ref| {
        object_ref.* = .{
            .abi_version = try reader.readU64(),
            .byte_length = try reader.readU64(),
            .sha256 = try reader.readDigest(),
        };
        try validateRef(object_ref.*);
    }
    const envelope_sha256 = try reader.readDigest();
    if (reader.position != encoded.len or !std.mem.eql(
        u8,
        &envelope_sha256,
        &envelopeSha256(encoded[0 .. encoded.len - digest_bytes]),
    )) return Error.InvalidEnvelope;
    return .{
        .config = config,
        .refs = refs,
        .envelope_sha256 = envelope_sha256,
    };
}

/// Verify the exact scalar identity and every typed external object.
pub fn decodeAndVerifyV1(
    encoded: []const u8,
    expected_config: ConfigV1,
    objects: ObjectsV1,
) Error!DecodedV1 {
    const decoded = try decodeManifestV1(encoded);
    if (!std.meta.eql(decoded.config, expected_config))
        return Error.InvalidComposition;
    for (object_kinds) |kind| {
        try validateObject(objects.get(kind));
        if (slicesOverlap(encoded, objects.get(kind).bytes))
            return Error.InvalidStorage;
    }

    var expected_storage: [encoded_bytes]u8 = undefined;
    const expected = try encodeV1(
        expected_config,
        objects,
        &expected_storage,
    );
    if (!std.mem.eql(u8, encoded, expected))
        return Error.InvalidComposition;
    return decoded;
}

pub fn objectRefV1(
    kind: ObjectKind,
    object: ObjectV1,
) Error!ObjectRefV1 {
    try validateObject(object);
    const byte_length = std.math.cast(u64, object.bytes.len) orelse
        return Error.InvalidObject;
    var hash = std.crypto.hash.sha2.Sha256.init(.{});
    hash.update(object_domain);
    hashU64(&hash, @intFromEnum(kind));
    hashU64(&hash, object.abi_version);
    hashU64(&hash, byte_length);
    hash.update(object.bytes);
    var sha256: Digest = undefined;
    hash.final(&sha256);
    return .{
        .abi_version = object.abi_version,
        .byte_length = byte_length,
        .sha256 = sha256,
    };
}

pub fn envelopeRootV1(encoded: []const u8) Error!Digest {
    return (try decodeManifestV1(encoded)).envelope_sha256;
}

fn validateConfig(config: ConfigV1) Error!void {
    if (config.execution_abi == 0 or
        config.request_epoch == 0 or
        config.publication_sequence == 0 or
        config.kv_tokens == 0 or
        config.output_tokens == 0 or
        config.output_tokens > config.kv_tokens or
        isZero(config.challenge_sha256))
        return Error.InvalidManifest;
    const has_parent = !isZero(config.parent_capsule_sha256);
    if ((config.checkpoint_generation == 0 and has_parent) or
        (config.checkpoint_generation != 0 and !has_parent))
        return Error.InvalidManifest;
}

fn validateObject(object: ObjectV1) Error!void {
    if (object.abi_version == 0 or object.bytes.len == 0)
        return Error.InvalidObject;
    _ = std.math.cast(u64, object.bytes.len) orelse
        return Error.InvalidObject;
}

fn validateRef(object_ref: ObjectRefV1) Error!void {
    if (object_ref.abi_version == 0 or
        object_ref.byte_length == 0 or
        isZero(object_ref.sha256))
        return Error.InvalidObject;
}

fn envelopeSha256(prefix: []const u8) Digest {
    var hash = std.crypto.hash.sha2.Sha256.init(.{});
    hash.update(envelope_domain);
    hash.update(prefix);
    var value: Digest = undefined;
    hash.final(&value);
    return value;
}

fn hashU64(hash: *std.crypto.hash.sha2.Sha256, value: u64) void {
    var bytes: [8]u8 = undefined;
    std.mem.writeInt(u64, &bytes, value, .little);
    hash.update(&bytes);
}

fn isZero(value: Digest) bool {
    return std.mem.allEqual(u8, &value, 0);
}

fn slicesOverlap(a: []const u8, b: []const u8) bool {
    if (a.len == 0 or b.len == 0) return false;
    const a_start = @intFromPtr(a.ptr);
    const b_start = @intFromPtr(b.ptr);
    const a_end = a_start + a.len;
    const b_end = b_start + b.len;
    return a_start < b_end and b_start < a_end;
}

const Writer = struct {
    bytes: []u8,
    position: usize = 0,

    fn writeBytes(self: *Writer, value: []const u8) Error!void {
        if (self.position > self.bytes.len or
            value.len > self.bytes.len - self.position)
            return Error.InvalidLength;
        @memcpy(self.bytes[self.position .. self.position + value.len], value);
        self.position += value.len;
    }

    fn writeU32(self: *Writer, value: u32) Error!void {
        var bytes: [4]u8 = undefined;
        std.mem.writeInt(u32, &bytes, value, .little);
        try self.writeBytes(&bytes);
    }

    fn writeU64(self: *Writer, value: u64) Error!void {
        var bytes: [8]u8 = undefined;
        std.mem.writeInt(u64, &bytes, value, .little);
        try self.writeBytes(&bytes);
    }

    fn writeDigest(self: *Writer, value: Digest) Error!void {
        try self.writeBytes(&value);
    }
};

const Reader = struct {
    bytes: []const u8,
    position: usize = 0,

    fn readBytes(self: *Reader, length: usize) Error![]const u8 {
        if (self.position > self.bytes.len or
            length > self.bytes.len - self.position)
            return Error.InvalidLength;
        const value = self.bytes[self.position .. self.position + length];
        self.position += length;
        return value;
    }

    fn readU32(self: *Reader) Error!u32 {
        var bytes: [4]u8 = undefined;
        @memcpy(&bytes, try self.readBytes(4));
        return std.mem.readInt(u32, &bytes, .little);
    }

    fn readU64(self: *Reader) Error!u64 {
        var bytes: [8]u8 = undefined;
        @memcpy(&bytes, try self.readBytes(8));
        return std.mem.readInt(u64, &bytes, .little);
    }

    fn readDigest(self: *Reader) Error!Digest {
        var value: Digest = undefined;
        @memcpy(&value, try self.readBytes(digest_bytes));
        return value;
    }
};

fn demoConfig() ConfigV1 {
    return .{
        .execution_abi = 0x4341_4558_0000_0001,
        .request_epoch = 0x4341_5251_0000_0001,
        .publication_sequence = 3,
        .checkpoint_generation = 0,
        .kv_tokens = 35,
        .output_tokens = 3,
        .challenge_sha256 = [_]u8{0xa7} ** 32,
    };
}

fn demoObjects() ObjectsV1 {
    return .{
        .model = .{ .abi_version = 0x4341_4d4f_0000_0001, .bytes = "model-v1:sha256:demo-glrt" },
        .tokenizer = .{ .abi_version = 0x4341_544b_0000_0001, .bytes = "tokenizer-v1:demo-qwen" },
        .execution_plan = .{ .abi_version = 0x4341_504c_0000_0001, .bytes = "plan-v1:cpu:threads=4:strict" },
        .resource_state = .{ .abi_version = 0x4341_5253_0000_0001, .bytes = "resource-v1:bank=17:kv=4096:output=64" },
        .lane_state = .{ .abi_version = 0x4341_4c4e_0000_0001, .bytes = "lane-v1:request=41:service=9" },
        .kv_state = .{ .abi_version = 0x4341_4b56_0000_0001, .bytes = "kv-v1:positions=35:root=demo" },
        .sampler_state = .{ .abi_version = 0x4341_534d_0000_0001, .bytes = "sampler-v1:rng=01020304:calls=3" },
        .output_state = .{ .abi_version = 0x4341_4f55_0000_0001, .bytes = "output-v1:tokens=901,902,903" },
        .publication_receipt = .{ .abi_version = 0x4341_5052_0000_0001, .bytes = "publication-v1:sequence=3:commit=demo" },
    };
}

test "continuation capsule layout is fixed" {
    try std.testing.expectEqual(@as(usize, 144), header_bytes);
    try std.testing.expectEqual(@as(usize, 48), object_ref_bytes);
    try std.testing.expectEqual(@as(usize, 9), object_count);
    try std.testing.expectEqual(@as(usize, 608), encoded_bytes);
}

test "continuation capsule verifies exact typed objects" {
    const config = demoConfig();
    const objects = demoObjects();
    var storage: [encoded_bytes]u8 = undefined;
    const encoded = try encodeV1(config, objects, &storage);
    const decoded = try decodeAndVerifyV1(encoded, config, objects);
    try std.testing.expectEqual(config, decoded.config);
    try std.testing.expectEqual(
        @as(u64, objects.kv_state.bytes.len),
        decoded.ref(.kv_state).byte_length,
    );
    try std.testing.expect(!isZero(decoded.envelope_sha256));
}

test "continuation capsule rejects every manifest mutation" {
    const config = demoConfig();
    const objects = demoObjects();
    var storage: [encoded_bytes]u8 = undefined;
    const encoded = try encodeV1(config, objects, &storage);
    var mutated: [encoded_bytes]u8 = undefined;
    for (0..encoded.len) |offset| {
        @memcpy(&mutated, encoded);
        mutated[offset] ^= 1;
        if (offset < encoded.len - digest_bytes) {
            const resealed = envelopeSha256(
                mutated[0 .. encoded.len - digest_bytes],
            );
            @memcpy(mutated[encoded.len - digest_bytes ..], &resealed);
        }
        if (decodeAndVerifyV1(&mutated, config, objects)) |_| {
            return error.ExpectedMutationRejection;
        } else |_| {}
    }
}

test "continuation capsule rejects valid foreign object substitution" {
    const config = demoConfig();
    const objects = demoObjects();
    var storage: [encoded_bytes]u8 = undefined;
    const encoded = try encodeV1(config, objects, &storage);
    var foreign = objects;
    foreign.kv_state = .{
        .abi_version = objects.kv_state.abi_version,
        .bytes = "kv-v1:positions=35:root=foreign",
    };
    try std.testing.expectError(
        Error.InvalidComposition,
        decodeAndVerifyV1(encoded, config, foreign),
    );
}

test "continuation capsule parent chain and storage fail closed" {
    const objects = demoObjects();
    var storage: [encoded_bytes]u8 = undefined;
    const first = try encodeV1(demoConfig(), objects, &storage);
    const parent = try envelopeRootV1(first);
    var resumed = demoConfig();
    resumed.publication_sequence = 4;
    resumed.checkpoint_generation = 1;
    resumed.kv_tokens = 36;
    resumed.output_tokens = 4;
    resumed.parent_capsule_sha256 = parent;
    var resumed_storage: [encoded_bytes]u8 = undefined;
    _ = try encodeV1(resumed, objects, &resumed_storage);

    var invalid = resumed;
    invalid.parent_capsule_sha256 = zero_digest;
    try std.testing.expectError(
        Error.InvalidManifest,
        encodeV1(invalid, objects, &resumed_storage),
    );

    var aliased = objects;
    aliased.model = .{
        .abi_version = objects.model.abi_version,
        .bytes = resumed_storage[0..16],
    };
    try std.testing.expectError(
        Error.InvalidStorage,
        encodeV1(resumed, aliased, &resumed_storage),
    );
}

File: bench/continuation_capsule.py
"""Independent codec and verifier for Glacier continuation capsule v1."""

from __future__ import annotations

import hashlib
import struct
from typing import Any


class CapsuleError(ValueError):
    """The continuation manifest or one of its bound objects is invalid."""


Digest = bytes
Record = dict[str, Any]
Object = tuple[int, bytes]

MAGIC = b"GCCAPV01"
WIRE_ABI = 0x4743434100000001
FLAG_REQUIRE_ALL_OBJECTS = 1
HEADER_BYTES = 144
OBJECT_REF_BYTES = 48
OBJECT_NAMES = (
    "model",
    "tokenizer",
    "execution_plan",
    "resource_state",
    "lane_state",
    "kv_state",
    "sampler_state",
    "output_state",
    "publication_receipt",
)
ENCODED_BYTES = HEADER_BYTES + len(OBJECT_NAMES) * OBJECT_REF_BYTES + 32
OBJECT_DOMAIN = b"glacier-continuation-object-v1\x00"
ENVELOPE_DOMAIN = b"glacier-continuation-capsule-wire-v1\x00"
ZERO_DIGEST = bytes(32)


def _u32(value: int) -> bytes:
    if not isinstance(value, int) or not 0 <= value <= 0xFFFFFFFF:
        raise CapsuleError("u32 out of range")
    return struct.pack("<I", value)


def _u64(value: int) -> bytes:
    if not isinstance(value, int) or not 0 <= value <= 0xFFFFFFFFFFFFFFFF:
        raise CapsuleError("u64 out of range")
    return struct.pack("<Q", value)


def _digest(value: bytes, *, allow_zero: bool = False) -> Digest:
    if not isinstance(value, bytes) or len(value) != 32:
        raise CapsuleError("invalid digest")
    if not allow_zero and value == ZERO_DIGEST:
        raise CapsuleError("zero digest")
    return value


def _hash(domain: bytes, *parts: bytes) -> Digest:
    hasher = hashlib.sha256()
    hasher.update(domain)
    for part in parts:
        hasher.update(part)
    return hasher.digest()


def _validate_config(config: Record) -> None:
    scalar_names = (
        "execution_abi",
        "request_epoch",
        "publication_sequence",
        "checkpoint_generation",
        "kv_tokens",
        "output_tokens",
    )
    for name in scalar_names:
        _u64(config[name])
    if (
        config["execution_abi"] == 0
        or config["request_epoch"] == 0
        or config["publication_sequence"] == 0
        or config["kv_tokens"] == 0
        or config["output_tokens"] == 0
        or config["output_tokens"] > config["kv_tokens"]
    ):
        raise CapsuleError("invalid scalar identity")
    _digest(config["challenge_sha256"])
    parent = _digest(config["parent_capsule_sha256"], allow_zero=True)
    has_parent = parent != ZERO_DIGEST
    if (config["checkpoint_generation"] == 0 and has_parent) or (
        config["checkpoint_generation"] != 0 and not has_parent
    ):
        raise CapsuleError("invalid parent chain")


def object_ref(kind_index: int, value: Object) -> Record:
    if not 0 <= kind_index < len(OBJECT_NAMES):
        raise CapsuleError("invalid object kind")
    abi_version, payload = value
    _u64(abi_version)
    if abi_version == 0 or not isinstance(payload, bytes) or not payload:
        raise CapsuleError("invalid object")
    byte_length = len(payload)
    _u64(byte_length)
    return {
        "abi_version": abi_version,
        "byte_length": byte_length,
        "sha256": _hash(
            OBJECT_DOMAIN,
            _u64(kind_index),
            _u64(abi_version),
            _u64(byte_length),
            payload,
        ),
    }


def encode(config: Record, objects: dict[str, Object]) -> bytes:
    """Encode a canonical fixed-size continuation manifest."""
    _validate_config(config)
    if set(objects) != set(OBJECT_NAMES):
        raise CapsuleError("object set mismatch")
    refs = [object_ref(index, objects[name]) for index, name in enumerate(OBJECT_NAMES)]
    prefix = b"".join(
        (
            MAGIC,
            _u64(WIRE_ABI),
            _u64(ENCODED_BYTES),
            _u32(FLAG_REQUIRE_ALL_OBJECTS),
            _u32(0),
            _u64(config["execution_abi"]),
            _u64(config["request_epoch"]),
            _u64(config["publication_sequence"]),
            _u64(config["checkpoint_generation"]),
            _u64(config["kv_tokens"]),
            _u64(config["output_tokens"]),
            config["challenge_sha256"],
            config["parent_capsule_sha256"],
            *(
                part
                for ref in refs
                for part in (
                    _u64(ref["abi_version"]),
                    _u64(ref["byte_length"]),
                    ref["sha256"],
                )
            ),
        )
    )
    if len(prefix) != ENCODED_BYTES - 32:
        raise CapsuleError("internal encoded length mismatch")
    return prefix + _hash(ENVELOPE_DOMAIN, prefix)


class _Reader:
    def __init__(self, data: bytes) -> None:
        self.data = data
        self.position = 0

    def take(self, length: int) -> bytes:
        end = self.position + length
        if length < 0 or end > len(self.data):
            raise CapsuleError("truncated capsule")
        value = self.data[self.position : end]
        self.position = end
        return value

    def u32(self) -> int:
        return struct.unpack("<I", self.take(4))[0]

    def u64(self) -> int:
        return struct.unpack("<Q", self.take(8))[0]

    def digest(self) -> Digest:
        return self.take(32)


def decode_manifest(encoded: bytes) -> Record:
    """Validate the self-contained envelope without external object authority."""
    if not isinstance(encoded, bytes) or len(encoded) != ENCODED_BYTES:
        raise CapsuleError("invalid capsule length")
    reader = _Reader(encoded)
    if reader.take(8) != MAGIC:
        raise CapsuleError("invalid magic")
    if reader.u64() != WIRE_ABI:
        raise CapsuleError("invalid ABI")
    if reader.u64() != ENCODED_BYTES:
        raise CapsuleError("invalid declared length")
    if reader.u32() != FLAG_REQUIRE_ALL_OBJECTS or reader.u32() != 0:
        raise CapsuleError("invalid flags")
    config = {
        "execution_abi": reader.u64(),
        "request_epoch": reader.u64(),
        "publication_sequence": reader.u64(),
        "checkpoint_generation": reader.u64(),
        "kv_tokens": reader.u64(),
        "output_tokens": reader.u64(),
        "challenge_sha256": reader.digest(),
        "parent_capsule_sha256": reader.digest(),
    }
    _validate_config(config)
    refs: dict[str, Record] = {}
    for name in OBJECT_NAMES:
        ref = {
            "abi_version": reader.u64(),
            "byte_length": reader.u64(),
            "sha256": reader.digest(),
        }
        if ref["abi_version"] == 0 or ref["byte_length"] == 0:
            raise CapsuleError("invalid object reference")
        _digest(ref["sha256"])
        refs[name] = ref
    envelope_sha256 = reader.digest()
    if reader.position != len(encoded):
        raise CapsuleError("trailing capsule bytes")
    expected_envelope = _hash(ENVELOPE_DOMAIN, encoded[:-32])
    if envelope_sha256 != expected_envelope:
        raise CapsuleError("envelope mismatch")
    return {
        "config": config,
        "refs": refs,
        "envelope_sha256": envelope_sha256,
    }


def decode_and_verify(
    encoded: bytes,
    expected_config: Record,
    objects: dict[str, Object],
) -> Record:
    """Verify the expected scalar identity and every exact external object."""
    decoded = decode_manifest(encoded)
    _validate_config(expected_config)
    if decoded["config"] != expected_config:
        raise CapsuleError("scalar identity substitution")
    if encode(expected_config, objects) != encoded:
        raise CapsuleError("object or manifest substitution")
    return decoded


def reseal_for_test(encoded: bytes) -> bytes:
    """Recompute only the outer envelope for mutation-completeness tests."""
    if len(encoded) != ENCODED_BYTES:
        raise CapsuleError("invalid capsule length")
    return encoded[:-32] + _hash(ENVELOPE_DOMAIN, encoded[:-32])


def demo_config() -> Record:
    return {
        "execution_abi": 0x4341455800000001,
        "request_epoch": 0x4341525100000001,
        "publication_sequence": 3,
        "checkpoint_generation": 0,
        "kv_tokens": 35,
        "output_tokens": 3,
        "challenge_sha256": bytes((0xA7,)) * 32,
        "parent_capsule_sha256": ZERO_DIGEST,
    }


def demo_objects() -> dict[str, Object]:
    return {
        "model": (0x43414D4F00000001, b"model-v1:sha256:demo-glrt"),
        "tokenizer": (0x4341544B00000001, b"tokenizer-v1:demo-qwen"),
        "execution_plan": (0x4341504C00000001, b"plan-v1:cpu:threads=4:strict"),
        "resource_state": (
            0x4341525300000001,
            b"resource-v1:bank=17:kv=4096:output=64",
        ),
        "lane_state": (0x43414C4E00000001, b"lane-v1:request=41:service=9"),
        "kv_state": (0x43414B5600000001, b"kv-v1:positions=35:root=demo"),
        "sampler_state": (
            0x4341534D00000001,
            b"sampler-v1:rng=01020304:calls=3",
        ),
        "output_state": (
            0x43414F5500000001,
            b"output-v1:tokens=901,902,903",
        ),
        "publication_receipt": (
            0x4341505200000001,
            b"publication-v1:sequence=3:commit=demo",
        ),
    }


def build_demo_bundle() -> Record:
    config = demo_config()
    objects = demo_objects()
    encoded = encode(config, objects)
    return {"config": config, "objects": objects, "encoded": encoded}
